### PR TITLE
EWPP-39: Fulltext - Facet widget

### PR DESCRIPTION
**Pull request by sergepavle**
### Description

[Insert description here]

### Change log

- Added:
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: tests/Kernel/FulltextWidgetTest.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Drupal\Tests\oe_list_pages\Kernel;
+
+use Drupal\oe_list_pages\ListSourceFactory;
+use Drupal\oe_list_pages\Plugin\facets\widget\ListPagesFulltextWidget;
+
+/**
+ * Text for Fulltext widget and query type.
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:9]: typo
```suggestion
 * Test for Fulltext widget and query type.
```

File: tests/Kernel/FulltextWidgetTest.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ */
+class FulltextWidgetTest extends ListsSourceBaseTest {
+
+  /**
+   * Creates a new processor object for use in the tests.
+   */
+  protected function setUp() {
+    parent::setUp();
+
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:18]: Unnecessary space.

File: tests/Kernel/FulltextWidgetTest.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
+<?php
+
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:2]: `declare(strict_types = 1);`

File: tests/Kernel/FulltextWidgetTest.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  protected function setUp() {
+    parent::setUp();
+
+    $this->widget = new ListPagesFulltextWidget(['fulltext_all_fields' => TRUE], 'oe_list_pages_fulltext', []);
+  }
+
+  /**
+   * Tests widget value conversion.
+   */
+  public function testWidgetFormValue() {
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:25]: missed `: void`

File: tests/Kernel/FulltextWidgetTest.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    $output = $this->widget->build($facet_inactive)[$facet_inactive->id()];
+    $this->assertSame('array', gettype($output));
+    $this->assertEquals('textfield', $output['#type']);
+    $this->assertEmpty($output['#default_value']);
+  }
+
+  /**
+   * Tests query type with and without fulltext in all fields.
+   */
+  public function testQueryType() {
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:46]: missed `: void`

File: tests/Kernel/FulltextWidgetTest.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    $query = $list->getQuery(0, 0, [], [$facet_name->id() => 'message']);
+    $query->execute();
+    $results = $query->getResults();
+    $this->assertCount(1, $results->getResultItems());
+  }
+
+  /**
+   * Tests query type.
+   */
+  public function testGetQueryType() {
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:78]: missed `: void`

File: tests/Kernel/FulltextWidgetTest.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    $result = $this->widget->getQueryType($this->queryTypes);
+    $this->assertEquals('fulltext_comparison', $result);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function testDefaultConfiguration() {
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:86]: missed `: void`

File: tests/Kernel/FulltextWidgetTest.php [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  /**
+   * Tests query type.
+   */
+  public function testGetQueryType() {
+    $result = $this->widget->getQueryType($this->queryTypes);
+    $this->assertEquals('fulltext_comparison', $result);
+  }
+
+  /**
+   * {@inheritdoc}
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:84]: `Test a default configuration.`

File: tests/Kernel/FulltextWidgetTest.php [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+   */
+  public function testQueryType() {
+    $this->createTestContent('item', 4);
+    // Another list for another bundle.
+    $item_list = $this->listFactory->get('entity_test_mulrev_changed', 'item');
+    $item_list->getIndex()->indexItems();
+
+    // Create facets for body and name.
+    $default_list_id = ListSourceFactory::generateFacetSourcePluginId('entity_test_mulrev_changed', 'item');
+    $facet = $this->createFacet('body', $default_list_id, '', 'oe_list_pages_fulltext', ['fulltext_all_fields' => TRUE]);
```
Comment (sergepavle) [inline tests/Kernel/FulltextWidgetTest.php:54]: Probably better `$facet_body`

File: src/Plugin/facets/query_type/Fulltext.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+      $field_identifier = $this->facet->getFieldIdentifier();
+      // Add the filter to the query if there are active values.
+      $active_items = $this->facet->getActiveItems();
+      $widget_config = $this->facet->getWidgetInstance()->getConfiguration();
+      foreach ($active_items as $value) {
+        $query->keys($value);
+        if (isset($widget_config['fulltext_all_fields']) && !$widget_config['fulltext_all_fields']) {
+          // Search on specific field.
+          $fulltext_fields = $query->getFulltextFields() ?? [];
+          $query->setFulltextFields([$field_identifier] + $fulltext_fields);
```
Comment (upchuk) [inline src/Plugin/facets/query_type/Fulltext.php:35]: probably shoiuld use array_merge here no? keyed by numbers they would not merge correctly?

File: src/Plugin/facets/widget/ListPagesFulltextWidget.php [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+   */
+  public function build(FacetInterface $facet) {
+    $build[$facet->id()] = [
+      '#type' => 'textfield',
+      '#title' => $facet->getName(),
+      '#default_value' => $this->getValueFromActiveFilters($facet, '0'),
+    ];
+
+    $build['#cache']['contexts'] = [
+      'url.query_args',
```
Comment (upchuk) [inline src/Plugin/facets/widget/ListPagesFulltextWidget.php:32]: Does this actually gets cached? It's a form..

File: src/Plugin/facets/widget/ListPagesFulltextWidget.php [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state, FacetInterface $facet) {
+    $form['fulltext_all_fields'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Fulltext search on all fields'),
+      '#default_value' => $this->getConfiguration()['fulltext_all_fields'],
```
Comment (upchuk) [inline src/Plugin/facets/widget/ListPagesFulltextWidget.php:61]: Let's add a #description:
```
Perform the search on all available indexed text fields.
```

File: src/Plugin/facets/widget/ListPagesWidgetInterface.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\facets\FacetInterface;
+
+/**
+ * Interface for list pages widget.
+ */
+interface ListPagesWidgetInterface {
+
+  /**
+   * Prepares the values to be passed to url generator using the submitted form.
```
Comment (upchuk) [inline src/Plugin/facets/widget/ListPagesWidgetInterface.php:16]: the URL generator..

File: src/Plugin/facets/widget/ListPagesWidgetInterface.php [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+   *
+   * @param \Drupal\facets\FacetInterface $facet
+   *   The facet.
+   * @param array $form
+   *   The form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state.
+   *
+   * @return array
+   *   The active filters to be handled by url generator.
```
Comment (upchuk) [inline src/Plugin/facets/widget/ListPagesWidgetInterface.php:26]: the URL generator

File: src/Plugin/facets/widget/ListPagesWidgetInterface.php [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+   */
+  public function prepareValueForUrl(FacetInterface $facet, array &$form, FormStateInterface $form_state): array;
+
+  /**
+   * Get active filters for the facet.
```
Comment (upchuk) [inline src/Plugin/facets/widget/ListPagesWidgetInterface.php:31]: the active filters...

File: tests/Kernel/FulltextWidgetTest.php [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+  public function testDefaultConfiguration(): void {
+    $default_config = $this->widget->defaultConfiguration();
+    $this->assertArrayHasKey('fulltext_all_fields', $default_config);
+    $this->assertTrue($default_config['fulltext_all_fields']);
+  }
+
+  /**
+   * Create test content.
+   */
+  protected function createTestContent($bundle, $count): void {
```
Comment (upchuk) [inline tests/Kernel/FulltextWidgetTest.php:103]: docblock and type hint

File: src/Plugin/facets/facet_source/ListFacetSource.php [hunk 17]
```diff
@@ -75,32 +83,11 @@ public static function create(ContainerInterface $container, array $configuratio
...
-      $container->get('entity_type.manager')
+      $container->get('entity_type.manager'),
+      $container->get('plugin.manager.search_api.display')
     );
   }
 
-  /**
-   * {@inheritdoc}
-   */
-  public function getFields() {
```
[the remark this answers is not in the document]
Comment (hernani) [reply src/Plugin/facets/facet_source/ListFacetSource.php:85]: Not needed in first place.

File: src/Plugin/facets/widget/ListPagesFulltextWidgetBase.php [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * The fulltext search widget.
+ *
+ * @FacetsWidget(
+ *   id = "oe_list_pages_fulltext",
+ *   label = @Translation("List pages fulltext"),
+ *   description = @Translation("A fulltext search widget."),
+ * )
+ */
+class ListPagesFulltextWidgetBase extends ListPagesWidgetBase {
```
Comment (upchuk) [inline src/Plugin/facets/widget/ListPagesFulltextWidgetBase.php:19]: This should not be a Base class.

**Pull request merged**